### PR TITLE
Upgrade Node.js to latest v12 LTS

### PR DESCRIPTION
**Pull request by craigthompsonhive**
[v12.16.2 Release Notes](https://github.com/nodejs/node/blob/master/doc/changelogs/CHANGELOG_V12.md#12.16.2)

### CODE REVIEW

**Comment by craigthompsonhive** [comment]
@googlebot I signed it!

**Comment by amio** [comment]
@craigthompsonhive Latest lts version is [12.14.0](https://github.com/nodejs/node/blob/master/doc/changelogs/CHANGELOG_V12.md#2019-12-17-version-12140-erbium-lts-mylesborins) now

**Comment by craigthompsonhive** [comment]
> @craigthompsonhive Latest lts version is [12.14.0](https://github.com/nodejs/node/blob/master/doc/changelogs/CHANGELOG_V12.md#2019-12-17-version-12140-erbium-lts-mylesborins) now

Updated :)

**Comment by amio** [comment]
ping @donmccasland @evanj 
Could you make a review on this? Many thanks 😘

**Comment by somethingnew2-0** [comment]
Bump: can this be reviewed and merged?

**Comment by evanj** [comment]
I am not actually an owner of this repository, so I can't actually review this. I also wonder if this major version upgrade is going to have the same issue we had with the base debian update: this may break people who are using this. Given that it is in the `experimental` directory maybe that doesn't matter?

**Comment by HofmannZ** [comment]
Latest is `12.16.1` now.

@evanj the bump from node 8 to node 10 was also not marked breaking. What's the approach for breaking upgrades in non-experimental images?

**Comment by evanj** [comment]
I think the answer is that it has not really been decided. The discussion at the time that I needed the debian 10 base image was to make it a new image name, to avoid breaking anyone depending on `latest` for the existing images.

**Comment by victiondev** [comment]
Is this ever going to be done?

**Comment by chanseokoh** [comment]
This repo has been primarily driven by the community; especially this will be true for node images, as the repo moderators don't have any node expertise. Should this ever moves forward, I think the consensus among the community will be the main driving force to evolve the node images in the ways however they see fit. Therefore, I can imagine I will be able to accept this PR, if, _for example_, the followings are true:

- A decent number of people with sufficient expertise in node, or the general community overall, agree that  it is a reasonable action to upgrade the nodejs version from v10 to v12 in Distroless.
- The consensus is that it does not warrant for creating and maintaining separate image repositories for v10 and v12 (e.g., `gcr.io/distroless/nodejs12` and `gcr.io/distroless/nodejs10`) or having additional v12 tags in the same repository (e.g., `gcr.io/distroless/nodejs:12` and `gcr.io/distroless/nodejs:10`).
- The community agrees that upgrading from v10 to v12 is an acceptable upgrade that does not introduce a lot of breakages and incompatibilities that make users using v10 unhappy.

If the community reaches on a consensus to merge this PR, I can certainly accept it.

**Comment by bishtawi** [comment]
Latest 12.x is now [12.16.2](https://github.com/nodejs/node/blob/master/doc/changelogs/CHANGELOG_V12.md#12.16.2). I would like to see this MR merged in sooner rather than later.

For me personally, I would like to upgrade to v12 asap, even at the cost of dropping v10 early. However, in the future, we should probably support distroless NodeJS versions for the last two LTS versions (different docker tag?) as we are just going to have the same issue when v14 LTS comes out.

File: WORKSPACE [hunk 1]
```diff
@@ -171,10 +171,10 @@ http_archive(
 http_archive(
     name = "nodejs",
     build_file = "//experimental/nodejs:BUILD.nodejs",
-    sha256 = "417bdc5402f6510fe1a5a898a9cdf1d67bd0202b5f014051c382f05358999534",
-    strip_prefix = "node-v10.17.0-linux-x64/",
+    sha256 = "52207f643ab0fba66d5189a51aac280c4834c81f24a7297446896386ec93a5ed",
```
Comment (craigthompsonhive) [inline WORKSPACE:174]:
```suggestion
    sha256 = "ffc92b8d9f53a10a2a734c93e78a3be74af7b0034bddd6f034a700c430eb94c1",
```

File: WORKSPACE [hunk 2]
```diff
@@ -171,10 +171,10 @@ http_archive(
...
+    strip_prefix = "node-v12.14.0-linux-x64/",
```
Comment (craigthompsonhive) [inline WORKSPACE:175]:
```suggestion
    strip_prefix = "node-v12.16.2-linux-x64/",
```

File: WORKSPACE [hunk 3]
```diff
@@ -171,10 +171,10 @@ http_archive(
...
     type = "tar.gz",
-    urls = ["https://nodejs.org/dist/v10.17.0/node-v10.17.0-linux-x64.tar.gz"],
+    urls = ["https://nodejs.org/download/release/latest-v12.x/node-v12.14.0-linux-x64.tar.gz"],
```
Comment (craigthompsonhive) [inline WORKSPACE:177]:
```suggestion
    urls = ["https://nodejs.org/download/release/v12.16.2/node-v12.16.2-linux-x64.tar.gz"],
```

File: examples/nodejs/Dockerfile [hunk 4]
```diff
@@ -1,4 +1,4 @@
-FROM node:10.17.0 AS build-env
+FROM node:12.14.0 AS build-env
```
Comment (craigthompsonhive) [inline examples/nodejs/Dockerfile:1]:
```suggestion
FROM node:12.16.2 AS build-env
```

File: examples/nodejs/node-express/Dockerfile [hunk 5]
```diff
@@ -1,4 +1,4 @@
-FROM node:10.17.0 AS build-env
+FROM node:12.14.0 AS build-env
```
Comment (craigthompsonhive) [inline examples/nodejs/node-express/Dockerfile:1]:
```suggestion
FROM node:12.16.2 AS build-env
```

**Comment by craigthompsonhive** [comment]
> This repo has been primarily driven by the community; especially this will be true for node images, as the repo moderators don't have any node expertise. Should this ever moves forward, I think the consensus among the community will be the main driving force to evolve the node images in the ways however they see fit. Therefore, I can imagine I will be able to accept this PR, if, _for example_, the following are true:
> 
> * A decent number of people with sufficient expertise in node, or the general community overall, agree that  it is a reasonable action to upgrade the nodejs version from v10 to v12 in Distroless.
> * The consensus is that it does not warrant for creating and maintaining separate image repositories for v10 and v12 (e.g., `gcr.io/distroless/nodejs12` and `gcr.io/distroless/nodejs10`) or having additional v10/v12 tags in the same repository (e.g., `gcr.io/distroless/nodejs:12` and `gcr.io/distroless/nodejs:10`).
> * The community agrees that upgrading from v10 to v12 is an acceptable upgrade that does not introduce a lot of breakages and incompatibilities that make users using v10 unhappy.
> * Or, it is a general consensus that v10 is mostly defunct and safe to drop without major concerns.
> 
> If the community reaches on a consensus to merge this PR, I can certainly accept it.

* I agree that different tags for different major versions of Node should be maintained, as Node.js tend to have multiple LTS/Latest versions released concurrently as per [their roadmap](https://nodejs.org/en/about/releases/).
* I think separate tags work fine personally, as this maps well to the [equivalent Docker images](https://hub.docker.com/_/node/)
* As per the roadmap, Node 10.x will be in LTS until `2021-04-30`. I think tagging as separate versions is the correct solution, rather than simply moving the current `node` image over to 12.x

**Comment by chanseokoh** [comment]
> Given the above, what is the process for introducing these tags @chanseokoh?

You simply need to create another `container_image`. Currently there are [two `container_image` rules](https://github.com/GoogleContainerTools/distroless/blob/3c6e10c21f508699c5ef7c84dee2ecc08bc8d5a3/experimental/nodejs/BUILD#L5-L13) (as iterated over the list `["", ":debug",]`). Note the `tars = ["@nodejs//:tar"]` is pointing to [the `nodejs` tar file](https://github.com/GoogleContainerTools/distroless/blob/3c6e10c21f508699c5ef7c84dee2ecc08bc8d5a3/WORKSPACE#L174) in WORKSPACE. So, you will need another `http_file` for v12 with a different name, create another `container_image` rule (plus `:debug` one), and refer to the v12 tar in the new rules.

Currently, `bazel run //experimental/nodejs:nodejs` and `bazel run //experimental/nodejs:debug` build the v10 images into your local Docker daemon. (`nodejs` and `debug` are the [`name` value](https://github.com/GoogleContainerTools/distroless/blob/3c6e10c21f508699c5ef7c84dee2ecc08bc8d5a3/experimental/nodejs/BUILD#L6) of the `container_image` rules.) So, when you have a new `container_image` rule with a name `nodejs12` for example, you'll do `bazel run //experimental/nodejs:nodejs12`.

Once we have a new `container_image` for v12, you can add new repository tags in [cloudbuild.yaml](https://github.com/GoogleContainerTools/distroless/blob/3c6e10c21f508699c5ef7c84dee2ecc08bc8d5a3/cloudbuild.yaml#L198-L206) (and this [BUILD](https://github.com/GoogleContainerTools/distroless/blob/3c6e10c21f508699c5ef7c84dee2ecc08bc8d5a3/BUILD#L68-L69)).

**Comment by LouisStAmour** [comment]
@chanseokoh Not sure if this is what you had in mind? This isn't a clean PR, but I copied the node folder and made basic edits to get node10, node12 and node14 folders: https://github.com/LouisStAmour/distroless/commit/cc0cf294f3f3a604830e4e3f88eff1ee764bf611 It appears to work locally within Docker for Mac using the shell script from that commit.

As an aside, I'll admit I really like this project but from a security and production-readiness perspective, I'm going to probably have to write a wrapper that can check for upstream updates and trigger changes to the WORKSPACE file and/or re-run builds when upstream Debian packages or upstream language runtimes change versions, to more consistently publish new versions and security-fix versions of images. This isn't unique to this project -- I've noticed that most Docker images, such as Microsoft's official "dotnet" images, will only apply security patches when they release new versions of their image, and they rarely if ever go back to update previous tags with new security fixes. Ideally, perhaps, there'd be a tool that could go through existing image layers and when new security fixes are available, create a replacement tar image layer using distroless, then scan a list of images to see which ones use the layer and create a new release by replacing the layer string in the previously published image. I wish more folks would end up releasing, let's say, Node v14.4.0 tagged as node:v14.4.0-0 aliased as v14.4.0 then when (inevitably) there's a security update in the static, base, nonroot or cc layers for whatever reason, automatically release node:v14.4.0-1 as the new v14.4.0 by replacing only the affected layer(s)...

**Comment by chanseokoh** [comment]
> Not sure if this is what you had in mind? This isn't a clean PR, but I copied the node folder and made basic edits to get node10, node12 and node14 folders: [LouisStAmour@cc0cf29](https://github.com/LouisStAmour/distroless/commit/cc0cf294f3f3a604830e4e3f88eff1ee764bf611) It appears to work locally within Docker for Mac using the shell script from that commit.

Great! However, although I no nothing about Node.js versions, I think we don't want to have separate directories for 10, 12, and 14? For example, we have a single Java directory publishing Java 8, 11, and future versions. Python 2 and 3 are the only exception, but I believe they [deserve the distinction](https://www.activestate.com/blog/python-3-vs-python-2-its-different-time/). (Python 3 is basically a different Python.)

> I'm going to probably have to write a wrapper that can check for upstream updates and trigger changes to the WORKSPACE file and/or re-run builds when upstream Debian packages or upstream language runtimes change versions, to more consistently publish new versions and security-fix versions of images.

That's awesome. This repo would also [love to have this monitoring system](https://github.com/GoogleContainerTools/distroless/issues/501#issuecomment-617545871) in place. :smile:

> for example static, base, nonroot or cc -- automatically release new images such as node:v14.4.0-1 as the new v14.4.0 by replacing only the affected layer(s)...

FYI, whenever there's a change to the repo srouce code (including changes to base images), Distroless images are automatically republished with the changes. It's just that it won't create a new tag like v14.4.0-1 every time.

**Comment by LouisStAmour** [comment]
I’ll see what I can prototype then, for both node versioning from the same sources and for automatic updates :) Given it is community-supported, I’ll admit I’m considering GitHub Actions as they’re free for open source. Google Cloud Build does not appear to have the same open source exemption listed on their pricing page that Microsoft supports via Azure DevOps and GitHub Actions...

**Comment by chanseokoh** [comment]
@LouisStAmour thanks a ton! I actually think GitHub Actions is a great idea.

**Comment by chanseokoh** [comment]
> I’ll see what I can prototype then, ... for automatic updates :) Given it is community-supported, I’ll admit I’m considering GitHub Actions as they’re free for open source.

@LouisStAmour in case you missed, automatic package updates is already [implemented](https://github.com/GoogleContainerTools/distroless/pull/544) using GitHub Actions. (Here's an example [PR](https://github.com/GoogleContainerTools/distroless/pull/549) created by the bot.)

**Comment by nhuray** [comment]
Do we plan to merge that PR any time soon ?

Node 12 is currently `12.18.2` and Node 14 is `14.5.0`

**Comment by chanseokoh** [comment]
Probably not. See https://github.com/GoogleContainerTools/distroless/pull/442#issuecomment-613402437

**Comment by michael-gillett** [comment]
I put up another PR to support building Node v12 and v14 along side v10. It should be easy to update when new major version come out in the future #556

**Comment by chanseokoh** [comment]
@michael-gillett that's cool! Thanks for making a contribution. I'll close this in favor for #556.

**Pull request closed without merging**